### ISSUE TITLE
An asp.net core web api application, using "UseWindowsService", reports an error “System.NotSupportedException: The content root changed. Changing the host configuration is not supported ” when starting the service 

### ISSUE THREAD

**Issue by quan0zhou** [dotnet/aspnetcore#36979]
- System 

  - Provider 

   [ Name]  .NET Runtime 
 
  - EventID 1026 

   [ Qualifiers]  0 
 
   Version 0 
 
   Level 2 
 
   Task 0 
 
   Opcode 0 
 
   Keywords 0x80000000000000 
 
  - TimeCreated 

   [ SystemTime]  2021-09-26T05:59:39.0432975Z 
 
   EventRecordID 14008 
 
   Correlation 
 
  - Execution 

   [ ProcessID]  0 
   [ ThreadID]  0 
 
   Channel Application 
 
   Computer USER-20210710BJ 
 
   Security 
 

- EventData 

   Application: ZhaoyangParking.exe CoreCLR Version: 6.0.21.45113 .NET Version: 6.0.0-rc.1.21451.13 Description: The process was terminated due to an unhandled exception. Exception Info: System.NotSupportedException: The content root changed. Changing the host configuration is not supported at Microsoft.AspNetCore.Builder.ConfigureHostBuilder.ConfigureHostConfiguration(Action`1 configureDelegate) at Microsoft.Extensions.Hosting.HostingHostBuilderExtensions.UseContentRoot(IHostBuilder , String ) at Microsoft.Extensions.Hosting.WindowsServiceLifetimeHostBuilderExtensions.UseWindowsService(IHostBuilder hostBuilder, Action`1 configure) at Microsoft.Extensions.Hosting.WindowsServiceLifetimeHostBuilderExtensions.UseWindowsService(IHostBuilder hostBuilder) at Program.<Main>$(String[] args) in E:\Learn\ZhaoyangParking\ZhaoyangParking\Program.cs:line 20  

**But There is nothing wrong with the program in .NET Version: 6.0.0-preview.7.21377.19**

**Comment by quan0zhou** [comment]
Find the relevant code through the source code
![image](https://user-images.githubusercontent.com/27860834/134798825-020bcf57-904a-4e3b-845c-5de4348900bf.png)
After decompiling the source code from Microsoft.Extensions.Hosting.WindowsServices.dll, I found the solution
![image](https://user-images.githubusercontent.com/27860834/134798941-fe5bb5e1-366f-4199-9b93-b7f0b6654231.png)

**Comment by quan0zhou** [comment]
`builder.Configuration.AddJsonFile(Path.Combine(AppContext.BaseDirectory, "appsettings.json"));

builder.Host
       //.UseWindowsService()//error: The content root changed. Changing the host configuration is not supported
       .ConfigureLogging(logging =>
       {
           logging.AddEventLog();

       }).ConfigureServices((hostContext, services)=>
       {
           services.AddSingleton<IHostLifetime, WindowsServiceLifetime>();
           services.Configure<EventLogSettings>(settings=>
           {
               if (string.IsNullOrEmpty(settings.SourceName))
               {
                   settings.SourceName = hostContext.HostingEnvironment.ApplicationName;
               }
           });
           services.AddHostedService<XXXXService>();


       });`

**Comment by davidfowl** [comment]
Yes this is unfortunate and one of the gotchas listed here (https://gist.github.com/davidfowl/0e0372c3c1d895c3ce195ba983b1e03d#differences-in-the-hosting-model). 

@quan0zhou Can you tell me what the content root is when you host as a windows service?

Also can you try this workaround:

```C#
using Microsoft.Extensions.Hosting.WindowsServices;

var options = new WebApplicationOptions
{
    Args = args,
    ContentRootPath = WindowsServiceHelpers.IsWindowsService() ? AppContext.BaseDirectory : default
};

var builder = WebApplication.CreateBuilder(options);

builder.Host.UseWindowsService();
```

**Comment by davidfowl** [comment]
> @quan0zhou Can you tell me what the content root is when you host as a windows service?

Don't worry, I have a repro with this. I think there might be an easy fix that makes this work by default.

**Comment by davidfowl** [comment]
@halter73 @Tratcher This is where we set the default content root to be the CWD https://github.com/dotnet/runtime/blob/dff742a06bb19fbb1a0ccfdefb14eb5379da3742/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L190, which is why the windows service logic needs to change it back to `AppContext.BaseDirectory`. If we could be a bit more clever about this default then it wouldn't require any explicit workarounds (for this case).

cc @eerhardt

**Comment by davidfowl** [comment]
I have another idea that doesn't require changing to much code. We update the documentation to pass the correct content root via command line args when creating the service:

```
sc config MyWebAppServiceTest binPath= "$pwd\WebApplication560.exe --contentRoot $pwd\"
```

We should update the docs to reflect this.

We also need to make a change to the host so that the content root comparison is on normalized paths. If you specify $pwd without the trailing slash, it fails because we compare the directories and it has to be an exact match today.

**Comment by davidfowl** [comment]
This should be a CTI scenario that we cover.

cc @captainsafia

**Comment by quan0zhou** [comment]
> 是的， 这是不幸的， 这里列出的哥查斯之一[（https://gist.github.com/davidfowl/0e0372c3c1d895c3ce195ba983b1e03d#differences-in-the-hosting-model）](https://gist.github.com/davidfowl/0e0372c3c1d895c3ce195ba983b1e03d#differences-in-the-hosting-model)。
> 
> @quan0zhou当您作为窗口服务托管时，你能告诉我内容根是什么吗？
> 
> 您也可以尝试此解决方法：
> 
> ```cs
> using Microsoft.Extensions.Hosting.WindowsServices;
> 
> var options = new WebApplicationOptions
> {
>     Args = args,
>     ContentRootPath = WindowsServiceHelpers.IsWindowsService() ? AppContext.BaseDirectory : default
> };
> 
> var builder = WebApplication.CreateBuilder(options);
> 
> builder.Host.UseWindowsService();
> ```

@davidfowl  This is a great solution, thanks!
